File: Niconicome/Models/Helper/Result/AttemptResult.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Niconicome.Models.Domain.Utils.Error;

namespace Niconicome.Models.Helper.Result
{
    public interface IAttemptResult
    {
        bool IsSucceeded { get; }
        string? Message { get; }
        Exception? Exception { get; }
        string ExceptionMessage { get; }
        ErrorLevel ErrorLevel { get; }
    }

    public class AttemptResult : IAttemptResult
    {
        public bool IsSucceeded { get; set; }

        public string? Message { get; set; }

        public Exception? Exception { get; set; }

        public ErrorLevel ErrorLevel { get; set; }

        public string ExceptionMessage => this.Exception?.Message ?? "None";

        /// <summary>
        /// インスタンスを作成
        /// </summary>
        /// <returns></returns>
        public static IAttemptResult Fail(string? message = null, Exception? ex = null, ErrorLevel errorLevel = ErrorLevel.Error)
        {
            return new AttemptResult() { Message = message, Exception = ex,ErrorLevel = errorLevel };
        }

        /// <summary>
        /// インスタンスを作成
        /// </summary>
        /// <param name="data"></param>
        /// <param name="message"></param>
        /// <returns></returns>
        public static IAttemptResult Succeeded(string? message = null)
        {
            return new AttemptResult() { IsSucceeded = true, Message = message };
        }
    }
}

File: NiconicomeWeb/src/shared/AttemptResult.ts
export interface AttemptResultWidthData<T> extends AttemptResult {
    
    /**
     * データ
     */
    readonly Data: T | null;
}

export interface AttemptResult {

    /**
     * 成功フラグ
     */
    readonly IsSucceeded: boolean;

    /**
     * メッセージ
     */
    readonly Message: string | null;
}


export class AttemptResultWidthDataImpl<T> implements AttemptResultWidthData<T> {

    constructor(isSucceeded: boolean, data: T | null, message: string | null) {
        this.IsSucceeded = isSucceeded;
        this.Data = data;
        this.Message = message;
    }

    readonly IsSucceeded: boolean;

    readonly Data: T | null;

    readonly Message: string | null;

    public static Succeeded<T>(data: T | null): AttemptResultWidthData<T> {
        return new AttemptResultWidthDataImpl(true, data, null);
    }

    public static Fail<T>(message: string): AttemptResultWidthData<T> {
        return new AttemptResultWidthDataImpl(false, null, message);
    }
}

export class AttemptResultImpl implements AttemptResult {

    constructor(isSucceeded: boolean, message: string | null) {
        this.IsSucceeded = isSucceeded;
        this.Message = message;
    }

    readonly IsSucceeded: boolean;

    readonly Message: string | null;

    public static Succeeded(): AttemptResult {
        return new AttemptResultImpl(true, null);
    }

    public static Fail(message: string): AttemptResult {
        return new AttemptResultImpl(false, message);
    }
}

